orangecontrib/bioinformatics/annotation/annotate_samples.py

```diff
@@ -0,0 +1,174 @@
+
+    Example of use:
+
+    >>> from Orange.data import Table
+    >>> from orangecontrib.bioinformatics.utils import serverfiles
+    >>> from orangecontrib.bioinformatics.annotation.annotate_samples import AnnotateSamples
+    >>>
+    >>> data = Table("https://datasets.orange.biolab.si/sc/aml-1k.tab.gz")
+    >>> markers_path = serverfiles.localpath_download(
+    >>>     'marker_genes','panglao_gene_markers.tab')
+    >>> marker = Table(markers_path)
+    >>> annotator = AnnotateSamples(p_value_th=0.05)
+    >>> annotations = annotator.annotate_samples(data, markers)
+
+    Attributes
+    ----------
+    p_value_th : float
+        A threshold for accepting the annotations. Annotations that has FDR
+        value bellow this threshold are used.
+    p_value_fun : callable, optional (defaults: statistics.Binomial().p_value)
+        A function that calculates p-value. It can be either
+        statistics.Binomial().p_value or hypergeom.sf.
+    """
+
+    def __init__(self, p_value_th, p_value_fun=statistics.Binomial().p_value):
+        if p_value_fun == hypergeom.sf:  # because sf accept x-1 instead of x
+            self.p_value_fun = lambda x, n, m, k: p_value_fun(x-1, n, m, k)
+        else:
+            self.p_value_fun = p_value_fun
+
+        self.p_threshold = p_value_th
+
+    @staticmethod
+    def _select_genes(data):
+        """
+        Function selects "over"-expressed attributes for items with Mann-Whitney
+        U test.
+
+        Parameters
+        ----------
+        data : Orange.data.Table
+            Tabular data
+
+        Returns
+        -------
+        :obj:`list`
+            Sets of selected attributes for each cell
+        """
+        if len(data.X) <= 1:
+            return [], []
+        # rank data
+        data_ge_ranked = rankdata(data.X, axis=0)
+
+        # compute U, mu, sigma
+        n = data_ge_ranked.shape[0]
+        n2 = n - 1
+        u = data_ge_ranked - 1
+        mu = n2 / 2
+        sigma = np.zeros(data_ge_ranked.shape[1])
+        for i in range(data_ge_ranked.shape[1]):
+            _, counts = np.unique(data_ge_ranked[:, i], return_counts=True)
+            sigma[i] = np.sqrt(
+                1 * n2 / 12 * ((n + 1) - np.sum((counts ** 3 - counts)) /
+                               (n * (n - 1))))
+
+        # compute z
+        z = (u - mu) / (sigma + 1e-16)
+
+        # gene selection
+        attributes_np = np.array([
+            a.attributes.get("Entrez ID") for a in data.domain.attributes])
+        attributes_sets = [set(attributes_np[row > 1]) - {None} for row in z]
+        return attributes_sets, z
+
+    @staticmethod
+    def _group_marker_attributes(markers):
+        """
+        Function transforms annotations table to dictionary with format
+        {annotation1: [attributes], annotation2: [attributes], ...}
+        """
+        types_dict = defaultdict(set)
+        for m in markers:
+            types_dict[str(m["Cell Type"])].add(int(m["Entrez ID"].value))
```

Comment:
"Entrez ID" can be unknown. Skip Entrez ID not present in the intersection between data table and marker table.

Reply:
Unknown Entrez IDs are already handled in line 87, where we subtract`{None}` from the set and thus remove all items that do not have an Entrez ID. Here it is not correct to get rid of markers which do not intersect with genes in the table since a number of marker genes for cell type are important information in p-value calculation.